core/src/main/java/org/opensearch/sql/datasource/DataSourceServiceImpl.java

```diff
@@ -50,57 +54,81 @@ public DataSourceServiceImpl(Set<DataSourceFactory> dataSourceFactories,
     this.dataSourceUserAuthorizationHelper = dataSourceUserAuthorizationHelper;
+    this.dataSourceLoaderCache = new DataSourceLoaderCacheImpl(dataSourceFactories);
   }
 
   @Override
-  public Set<DataSourceMetadata> getDataSourceMetadataSet() {
+  public Set<DataSourceMetadata> getDataSourceMetadataSet(boolean isDefaultDataSourceRequired) {
     List<DataSourceMetadata> dataSourceMetadataList
         = this.dataSourceMetadataStorage.getDataSourceMetadata();
     Set<DataSourceMetadata> dataSourceMetadataSet = new HashSet<>(dataSourceMetadataList);
-    dataSourceMetadataSet.add(DataSourceMetadata.defaultOpenSearchDataSourceMetadata());
+    if (isDefaultDataSourceRequired) {
+      dataSourceMetadataSet.add(DataSourceMetadata.defaultOpenSearchDataSourceMetadata());
+    }
+    removeAuthInfo(dataSourceMetadataSet);
     return dataSourceMetadataSet;
   }
 
+  @Override
+  public DataSourceMetadata getDataSourceMetadataSet(String datasourceName) {
+    Optional<DataSourceMetadata> dataSourceMetadataOptional
+        = this.dataSourceMetadataStorage.getDataSourceMetadata(datasourceName);
+    if (dataSourceMetadataOptional.isEmpty()) {
+      throw new IllegalArgumentException("DataSource with name: " + datasourceName
+          + " doesn't exist.");
+    }
+    removeAuthInfo(dataSourceMetadataOptional.get());
+    return dataSourceMetadataOptional.get();
+  }
+
 
   @Override
   public DataSource getDataSource(String dataSourceName) {
     Optional<DataSourceMetadata>
         dataSourceMetadataOptional = getDataSourceMetadata(dataSourceName);
     if (dataSourceMetadataOptional.isEmpty()) {
-      throw new IllegalArgumentException(
+      throw new DataSourceNotFoundException(
           String.format("DataSource with name %s doesn't exist.", dataSourceName));
     } else {
       DataSourceMetadata dataSourceMetadata = dataSourceMetadataOptional.get();
       this.dataSourceUserAuthorizationHelper
           .authorizeDataSource(dataSourceMetadata);
-      return getDataSourceFromMetadata(dataSourceMetadata);
+      return dataSourceLoaderCache.getOrLoadDataSource(dataSourceMetadata);
     }
   }
 
   @Override
   public void createDataSource(DataSourceMetadata metadata) {
     validateDataSourceMetaData(metadata);
     if (!metadata.getName().equals(DEFAULT_DATASOURCE_NAME)) {
+      this.dataSourceLoaderCache.getOrLoadDataSource(metadata);
       this.dataSourceMetadataStorage.createDataSourceMetadata(metadata);
     }
-    dataSourceMap.put(metadata,
-        dataSourceFactoryMap.get(metadata.getConnector()).createDataSource(metadata));
   }
 
   @Override
   public void updateDataSource(DataSourceMetadata dataSourceMetadata) {
-    throw new UnsupportedOperationException("will be supported in future");
+    validateDataSourceMetaData(dataSourceMetadata);
+    if (!dataSourceMetadata.getName().equals(DEFAULT_DATASOURCE_NAME)) {
+      this.dataSourceLoaderCache.getOrLoadDataSource(dataSourceMetadata);
```

Comment:
Everytime I need datasource or get metadata, I query it from index, so it will always be the latest. If the queried metadata is same as the last one, it will use cached Datasource object or otherwise it will create a new one.